File: Data Structures & Algorithms/count-paths/submission-5.java
class Solution {

    // grid[row][col] = grid[row-1][col] + grid[row][col-1];
    // 如果只有一個一維陣列，從左邊往右邊掃的話，grid[col] => 這個代表上一個 => grid[row-1][col]
    // 我們可以發現一個規律 grid[col] = grid[col] + grid[col - 1] => grid[col] += grid[col - 1]
     public int uniquePaths(int m, int n) {
        int[] f = new int[n];
        Arrays.fill(f, 1);

       for (int row = 1; row < m; row++) {
        for (int col = 1; col < n; col++) {
           f[col] += f[col - 1];
        }
       } 
      
      return f[n - 1];
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dirs = [(0, 1), (0, -1), (-1, 0), (1, 0)]
        m = len(grid)
        n = len(grid[0])

        def dfs(grid, row, col):
            if 0 <= row < m and 0 <= col < n and grid[row][col] != '0':   
                grid[row][col] = '0'
                for dr, dc in dirs:
                    new_row, new_col = row + dr, col + dc
                    dfs(grid, new_row, new_col)    

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    dfs(grid, i, j)
                    count += 1
        return count            
    